docs/_posts/2023-08-25-foreign-function-interface.markdown

```diff
@@ -0,0 +1,271 @@
+
+![Plot](./jmh/plot/jmh-result-pinnable-vs-output-plot.pdf)
+
+- `ffiGetOutputSlice()` is faster when values read are  1k in size or smaller. The cost of an extra copy in the C++ side from the pinnable slice buffer into the supplied buffer allocated by Java Foreign Memory API is less than the cost of the extra call to release a pinnable slice.
+- `ffiGetPinnableSlice()` is faster when values read are 4k in size, or larger. Consistent with intuition, the advantage grows with larger read values.
+
+The way that the RocksDB API is constructed means that of the 2 methods compared, `ffiGetOutputSlice()` will always make exactly 1 more copy than `ffiGetPinnableSlice()`. The underlying RocksDB C++ API will always copy into its own temporary buffer if it decides that it cannot pin an internal buffer, and that will be returned as the pinnable slice. There is a potential optimization where the temporary buffer could be replaced by an output buffer, such as that supplied by `ffiGetOutputSlice()`; in practice that is a hard fix to hack in. Its effectiveness depends on how often RocksDB fails to pin an internal buffer.
+
+A solution which either filled a buffer *or* returned a pinnable slice would give us the best of both worlds.
+
+## Other Conclusions
+
+### Build Processing
+
+- It is easier to implement an interface using FFI than JNI. No intermediate build processing or code generation steps were needed to implement this protoype.
+
+- For a production version, we would urge using `jextract` to automate the process of generating Java API methods from the set of supporting stubs we generate.
+
+### Safety
+
+- The use of `jextract` will give a similar level of type security to the use of JNI, when crossing the language boundary. But we do not believe FFI is significantly more type-safe than JNI for method invocation. Neither is it less safe, though.
+
+### Native Memory
+
+Panama's *Foreign-Memory Access API* appears to us to be the most significant part of the whole project. At the `Java` side of RocksDB it gives us a clean mechanism (a `MemorySegment`) for holding RocksDB data (e.g. as from the result of a `get()`) call pending its forwarding to client code or network buffers.
+
+We have taken advantage of this mechanism to provide the core `FFIDB.getPinnableSlice()` method in our Panama-based API. The rest of our prototype `get()` API, duplicating the existing *JNI*-based API, is then a *Pure Java* library on top of `FFIDB.getPinnableSlice()` and `FFIPinnableSlice.reset()`.
+
+The common standard for foreign memory opens up the possibility of efficient interoperation between RocksDB and Java clients (e.g. Kafka). We think that this is really the key to higher performing, more integrated Java-based systems:
+- This could result in data never being copied into Java memory, or a significant reduction in copies, as native `MemorySegment`s are handed off between co-operating Java clients of fundamentally native APIs.
```

Comment:
is it more error prone? Does the client need something higher level than `MemorySegment` so they can call a cleanup function like `PinnableSlice::reset()`?

Reply:
The wrapper-ised `get()` variant does this. I have revised the text to be a bit clearer. But it falls out as more of a layered approach, 2-step with `getPinnableSlice()` and `reset()` and the potential for chaining performance, or wrapped with `get()` and other high level methods.